dsgrid/utils/dataset.py

```diff
@@ -208,3 +209,56 @@ def remove_invalid_null_timestamps(df, time_columns, stacked_columns):
         .filter(f"{time_column} IS NOT NULL or count_time == 0")
         .select(orig_columns)
     )
+
+
+@track_timing(timer_stats_collector)
+def repartition_if_needed_by_mapping(
+    df: DataFrame,
+    value_columns: list[str],
+    mapping_type: DimensionMappingType,
+    dimension_type: DimensionType,
+    scratch_dir_context: ScratchDirContext,
+) -> DataFrame:
+    """Repartition the dataframe if the mapping will cause data skew."""
+    # We experienced an issue with the DECARB buildings dataset where the disaggregation of
+    # region to county caused a major issue where one Spark executor thread got stuck,
+    # seemingly indefinitely. A message like this was repeated continually.
+    # UnsafeExternalSorter: Thread 152 spilling sort data of 4.0 GiB to disk (0  time so far)
+    # It appears to be caused by data skew, though the imbalance didn't seem too severe.
+    # Using a variation of what online sources call a "salting technique" solves the issue.
+    # Apply the technique to mappings that will cause an explosion of rows.
+    # Note that this probably isn't needed in all cases and we may need to adjust in the
+    # future.
+    # The case with buildings was particularly severe because it is an unpivoted dataset.
+
+    # Note: log messages below are checked in the tests.
+    if mapping_type in {
+        DimensionMappingType.ONE_TO_MANY_DISAGGREGATION,
+        DimensionMappingType.ONE_TO_MANY_ASSIGNMENT,
```

Comment:
Based on team meeting: Only do this for ONE_TO_MANY_DISAGGREGATION. And add a column of random floats.